### PR TITLE
Backwards compatibility for tool/clippy lints

### PR DESCRIPTION
**Pull request by flip1995**
cc #44690 
cc https://github.com/rust-lang-nursery/rust-clippy/pull/2977#issuecomment-409706557

This is the next step towards `tool_lints`.

This makes Clippy lints still work without scoping, but will warn and suggest the new scoped name. This warning will only appear if the code is checked with Clippy itself.

There is still an issue with using the old lint name in inner attributes. For inner attributes the warning gets emitted twice. I'm currently not really sure why this happens, but will try to fix this ASAP.

r? @Manishearth

### CODE REVIEW

File: src/test/ui-fulldeps/lint_tool_test.stderr
Comment (flip1995) [inline src/test/ui-fulldeps/lint_tool_test.stderr]: This warning is emitted twice

**Review by Manishearth — APPROVED** [review]
design LGTM. This is somewhat temporary code so I'm not looking too closely.

I can r+, but perhaps we should r? @nrc

**Comment by Manishearth** [comment]
Did a closer pass through, looks good.

Might be good to also lint on `cfg_attr(clippy)` asking people to remove it in favor of these but that's trickier to do. Perhaps just mention in the lint warning that cfg(clippy) is also not necessary anymore.

r=manishearth it if you end up making those changes and/or feel it's otherwise ready

@bors delegate+

**Comment by Manishearth** [comment]
btw, feel free to submodule your clippy PR here so we don't have to do the submodule dance, just be careful not to push to that PR once this merges (I'll merge it on that side)

**Comment by Manishearth** [comment]
We can also merge into a staging branch if you'd like

**Comment by flip1995** [comment]
> Might be good to also lint on cfg_attr(clippy) asking people to remove it in favor of these but that's trickier to do.

I tried to write a Clippy lint for that, but I couldn't get it to lint on inner attributes. (unfinished: https://github.com/flip1995/rust-clippy/commits/cfg_attr_lint) I think the problem was, that they never appear in the AST. Either they were already removed or converted to just `allow/warn/...`. I would add a hint, that also `cfg_attr` are deprecated.

Before merging this I'd like to fix https://github.com/rust-lang/rust/pull/53762#discussion_r213297975

> btw, feel free to submodule your clippy PR here

How can I do that? My git-fu isn't good enough for that. 😄

**Comment by Manishearth** [comment]
Is the clippy side PR ready? Give me push access (tick the box on this page that lets rust collab push to the PR) and I'll do the rest.

**Comment by flip1995** [comment]
The Clippy PR is ready. It builds without errors with this PR. The box is ticked, thanks for doing this!

**Comment by Manishearth** [comment]
Gonna wait for https://github.com/rust-lang/rust/pull/53758 to land first, but will do soon!

File: src/librustc/lint/context.rs
Comment (flip1995) [inline src/librustc/lint/context.rs]: FIXME: This should be `Err((Some(ids), complete_name)`

**Comment by Manishearth** [comment]
Unsure why that's unexpected, it's included in the error.

**Comment by flip1995** [comment]
I couldn't find the reason for the double emitted warning on inner attributes today... I'm not sure what I'm doing different than in the UNKNOWN_LINTS case, which does not get linted twice.

**Comment by Manishearth** [comment]
Ah. Could you push a commit updating test expectations and r=me then? We
can fix this later

On Aug 29, 2018 12:39 PM, "Philipp Krones" <notifications@github.com> wrote:

I couldn't find the reason for the double emitted warning on inner
attributes today... I'm not sure what I'm doing different than in the
UNKNOWN_LINTS case, which does not get linted twice.

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/rust-lang/rust/pull/53762#issuecomment-417079086>, or mute
the thread

**Comment by nrc** [comment]
lgtm

**Comment by flip1995** [comment]
Just added another test for `cfg_attr`, while fixing the compile-fail-fulldeps test. It seems, that only the first inner attribute gets linted twice. hmm..

@bors r=Manishearth

**Comment by flip1995** [comment]
This error is caused by rust-lang-nursery/rust-clippy#3104. What needs to change is the import of `declare_lint`->`declare_tool_lint` and a `*.stderr` file update.

What would be the best way to do this? Pushing to rust-lang-nursery/rust-clippy#2977 and pointing this PR to the new commit?

**Comment by Manishearth** [comment]
Yes, please push an update, I'll do the rest.

**Comment by Manishearth** [comment]
Oh, it's a simple fix, I can do it

**Comment by Manishearth** [comment]
Done 

@bors r+

**Comment by flip1995** [comment]
Thanks!

**Comment by Arnavion** [comment]
What does "lint name `{}` is deprecated and may not have an effect in the future. Also `cfg_attr(cargo-clippy)` won't be necessary anymore" mean by the second sentence? The way it's worded makes it sound like I still need to keep the `cfg_attr`, but then why is it warning me about it now? If I'm supposed to remove it now, why does it say "*won't* be necessary" ?

**Comment by flip1995** [comment]
It isn't necessary on nightly, with `tool_lints` feature turned on. But it is still necessary on stable, until `tool_lints` are stable (see rust-lang-nursery/rust-clippy#3159). Maybe the `cfg_attr` note was a bit premature.

**Comment by Arnavion** [comment]
Thanks for the link to that issue.

File: src/test/ui-fulldeps/lint_tool_test.stderr
Comment (flip1995) [inline src/test/ui-fulldeps/lint_tool_test.stderr]: The reason that this warning is emitted twice is the note `#[warn(renamed_and_removed_lints)] on by default`, which only appears on the first occurrence of lint warnings. That doesn't effect the `UNKNOWN_LINTS` lint though.. :thinking:

**Pull request merged**
